File: bench.tables.py
#!/usr/bin/env python3
"""Time read/write of t/HepatitisCdata.csv over 10 runs each.

read_table  -> pandas.read_csv   (csv-parsing form; his read_table -> _parse_csv_file)
write_table -> DataFrame.to_csv

Reads t/HepatitisCdata.csv; writes the same data to a temp file in /tmp.
Emits python.read_table.tsv and python.write_table.tsv, one row per run.
Optional argv[1] overrides the input path.
"""
import os
import sys
import tempfile
import time

import pandas as pd


def save_times(fname, times):
    with open(fname, 'w') as fh:
        fh.write('run\tseconds\n')
        for i, t in enumerate(times):
            fh.write(f'{i}\t{t:.9f}\n')


def main():
    src = sys.argv[1] if len(sys.argv) > 1 else 't/HepatitisCdata.csv'
    d = pd.read_csv(src)

    tmp = tempfile.NamedTemporaryFile(mode='w', suffix='.csv', dir='/tmp', delete=False)
    tmp.close()
    path = tmp.name

    write = []
    for _ in range(10):
        t0 = time.perf_counter()
        d.to_csv(path, index=False)
        t1 = time.perf_counter()
        write.append(t1 - t0)
    save_times('python.write_table.tsv', write)

    read = []
    for _ in range(10):
        t0 = time.perf_counter()
        _ = pd.read_csv(src)
        t1 = time.perf_counter()
        read.append(t1 - t0)
    save_times('python.read_table.tsv', read)

    os.unlink(path)


if __name__ == '__main__':
    main()

File: bench.tables.R
#!/usr/bin/env Rscript
# Time read/write of t/HepatitisCdata.csv over 10 runs each.
#   read_table  -> read.csv                     (his read_table -> _parse_csv_file)
#   write_table -> write.csv / write.table(sep = ",")
# Reads t/HepatitisCdata.csv; writes the same data to a temp file in /tmp.
# Emits R.read_table.tsv and R.write_table.tsv, one row per run.
# Optional first arg overrides the input path.

save_times <- function(fname, times) {
    out <- data.frame(run = seq_along(times) - 1L, seconds = times)
    write.table(out, fname, sep = "\t", quote = FALSE, row.names = FALSE)
}

main <- function() {
    args <- commandArgs(trailingOnly = TRUE)
    src <- if (length(args) > 0) args[1] else "t/HepatitisCdata.csv"
    d <- read.csv(src, stringsAsFactors = FALSE)
    path <- tempfile(tmpdir = "/tmp", fileext = ".csv")

    write_t <- numeric(10)
    for (i in seq_len(10)) {
        t0 <- Sys.time()
        write.csv(d, path, row.names = FALSE)
        t1 <- Sys.time()
        write_t[i] <- as.numeric(t1) - as.numeric(t0)
    }
    save_times("R.write_table.tsv", write_t)

    read_t <- numeric(10)
    for (i in seq_len(10)) {
        t0 <- Sys.time()
        invisible(read.csv(src, stringsAsFactors = FALSE))
        t1 <- Sys.time()
        read_t[i] <- as.numeric(t1) - as.numeric(t0)
    }
    save_times("R.read_table.tsv", read_t)

    unlink(path)
}

main()
